### PR TITLE
Flattened Contract Registry + Three Agents Integration

### PR DESCRIPTION
**Pull request by KPrasch**
Refactors the ethereum contract registrar to store a single list of contract data, while giving contract agents the ability to lookup contract components by name, and assemble a unified proxy contract instance by reading live dispatcher target values from the blockchain.

This PR also includes PR #310 and integrates test modifications from PR #318

### CODE REVIEW

File: nucypher/blockchain/eth/interfaces.py [hunk 1]
```diff
@@ -323,17 +274,79 @@ def get_contract_factory(self, contract_name) -> Contract:
 
         return contract
 
-    def get_contract_address(self, contract_name: str) -> List[str]:
-        """Retrieve all known addresses for this contract"""
-        contracts = self._registrar.lookup_contract(contract_name=contract_name)
-        addresses = [c['addr'] for c in contracts]
-        return addresses
+    def __wrap_contract(self, dispatcher_contract: Contract,
+                        target_contract: Contract, factory=Contract) -> Contract:
```
Comment (szotov) [inline nucypher/blockchain/eth/interfaces.py]: Is this the same method in the deployers.py (103 line)?

File: tests/blockchain/eth/interfaces/test_registry.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def test_contract_registrary(tempfile_path):
+
+    with pytest.raises(EthereumContractRegistry.RegistryError):
+        bad_registrar = EthereumContractRegistry(registry_filepath='/fake/file/path/registry.json')
+        bad_registrar.search(contract_address='0xdeadbeef')
+
+    # Tests everything is as it should be when initially created
+    test_registrar = EthereumContractRegistry(registry_filepath=tempfile_path)
```
Comment (szotov) [inline tests/blockchain/eth/interfaces/test_registry.py]: Maybe rename to test_registry?

File: tests/blockchain/eth/utilities.py [hunk 3]
```diff
@@ -75,58 +71,13 @@ def generate_accounts(w3: Web3, quantity: int) -> List[str]:
     return addresses
 
 
-class TesterPyEVMBackend(PyEVMBackend):
```
Comment (tuxxy) [inline tests/blockchain/eth/utilities.py]: 👍

File: tests/utilities.py [hunk 4]
```diff
@@ -24,34 +24,59 @@
...
-        ursula.attach_rest_server()
-
-        class MockDatastoreThreadPool(object):
-            def callInThread(self, f, *args, **kwargs):
-                return f(*args, **kwargs)
-
-        ursula.datastore_threadpool = MockDatastoreThreadPool()
-        ursula.dht_listen()
+
+        if bare:
```
Comment (tuxxy) [inline tests/utilities.py]: Can you make a comment on what the context of `bare` is here?

File: tests/blockchain/eth/interfaces/test_registry.py [hunk 5]
```diff
@@ -3,29 +3,29 @@
 from nucypher.blockchain.eth.interfaces import EthereumContractRegistry
 
 
-def test_contract_registrary(tempfile_path):
+def test_contract_registryy(tempfile_path):
```
Comment (michwill) [inline tests/blockchain/eth/interfaces/test_registry.py]: Eh? Why `registryyyy`?

File: tests/utilities.py (hunk 4 above)
Comment (michwill) [reply tests/utilities.py]: I think, this is a "script kiddie" (c) Ursula :scream:

File: nucypher/blockchain/eth/agents.py [hunk 6]
```diff
@@ -61,7 +61,8 @@ def get_balance(self, address: str=None) -> int:
 
 
 class NucypherTokenAgent(EthereumContractAgent):
-    _principal_contract_name = "NuCypherToken"
+    principal_contract_name = "NuCypherToken"
+    _upgradeable = False
```
Comment (michwill) [inline nucypher/blockchain/eth/agents.py]: I really like the `_upgradeable` being here :-)

Comment (KPrasch) [reply nucypher/blockchain/eth/agents.py]: :heavy_check_mark:

File: tests/utilities.py (hunk 4 above)
Comment (KPrasch) [reply tests/utilities.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/interfaces.py [hunk 7]
```diff
@@ -327,17 +278,79 @@ def get_contract_factory(self, contract_name) -> Contract:
...
+
+                if len(matching_pairs) == 0:
+                    raise self.InterfaceError("No dispatcher targets known contract records for {}".format(name))
+
+                elif len(matching_pairs) > 1:
+                    raise self.InterfaceError("There is more than one dispatcher targeting {}".format(name))
+
+                selected_contract_address, selected_contract_abi = matching_pairs[0]
+        else:
+            if len(target_contract_records) > 1:  # TODO: Allow multiple non-upgradeable records (UserEscrow)
```
Comment (jMyles) [inline nucypher/blockchain/eth/interfaces.py]: Supposed to be an exactly-one pattern.

File: nucypher/blockchain/eth/agents.py [hunk 8]
```diff
@@ -15,29 +13,31 @@ class EthereumContractAgent(ABC):
...
+    def __init__(self, blockchain: Blockchain=None, *args, **kwargs):
 
         if blockchain is None:
             blockchain = Blockchain.connect()
         self.blockchain = blockchain
 
-        if contract is None:
-            address = blockchain.interface.get_contract_address(contract_name=self._principal_contract_name)[-1]  # TODO: Handle multiple
-            contract = blockchain.interface.get_contract(address)
+        # Fetch the contract by reading address and abo from the registry and blockchain
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/agents.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/interfaces.py (hunk 7 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/interfaces.py]: :heavy_check_mark:

**Pull request merged**